commit 8573b8b8be9b4dec215f5b63128e2a26f14b94cf (refs/changes/09/5509/3)
Author: Trent Mick <trentm@gmail.com>
Date:   2019-02-08T16:51:42-08:00 (8 months ago)
    
    joyent/node-haproxy-log#1 reported malformed log line for entry from log file

diff --git a/CHANGES.md b/CHANGES.md
index 7e0abeb..849c9f4 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,14 @@
 
 No changes.
 
+## v1.0.1
+
+- Issue #1: Fix parsing of some edge cases:
+    - possible `-1` for `status_code`
+    - possible `-1` for some fields rather than the looser "possibly
+      negative number"
+    - possible `+` on the `Tt` and `retries` fields
+
 ## v1.0.0
 
 Initial release.
diff --git a/lib/haproxy_log_parser.js b/lib/haproxy_log_parser.js
index 9178b69..0fa81a9 100644
--- a/lib/haproxy_log_parser.js
+++ b/lib/haproxy_log_parser.js
@@ -11,7 +11,7 @@ var VE = require('verror');
 var CLS_NOT_SPACE = '[^ ]+';
 var CLS_DIGITS = '[0-9]+';
 var CLS_IPADDR = '[:a-f0-9\\.]+';
-var CLS_DIGITS_NEG = '[-0-9]+';
+var CLS_DIGITS_OR_MINUS_ONE = '[0-9]+|-1';
 
 function
 build_matcher()
@@ -28,16 +28,6 @@ build_matcher()
 		re_str += re;
 	};
 
-	var numseq = function () {
-		for (var i = 0; i < arguments.length; i++) {
-			if (i > 0) {
-				non('/');
-			}
-
-			group(CLS_DIGITS_NEG, arguments[i], 'number');
-		}
-	};
-
 	/*
 	 * Construct a regular expression that matches the basic HAProxy HTTP
 	 * log format without any of the optional fields.
@@ -60,18 +50,43 @@ build_matcher()
 	group('[^ /]+', 'backend_name');
 	non('/');
 	group(CLS_NOT_SPACE, 'server_name');
+
 	non(' ');
-	numseq('Tq', 'Tw', 'Tc', 'Tr', 'Tt');
+	group(CLS_DIGITS_OR_MINUS_ONE, 'Tq', 'number');
+	non('/');
+	group(CLS_DIGITS_OR_MINUS_ONE, 'Tw', 'number');
+	non('/');
+	group(CLS_DIGITS_OR_MINUS_ONE, 'Tc', 'number');
+	non('/');
+	group(CLS_DIGITS_OR_MINUS_ONE, 'Tr', 'number');
+	non('/');
+	group('\\+?', 'Tt_logasap', 'notempty');
+	group(CLS_DIGITS, 'Tt', 'number');
+
 	non(' ');
-	group(CLS_DIGITS, 'status_code', 'number');
+	group(CLS_DIGITS_OR_MINUS_ONE, 'status_code', 'number');
 	non(' ');
 	group(CLS_DIGITS, 'bytes_read', 'number');
 	non(' - - ');
 	group('....', 'termination_state', 'termination_state');
+
 	non(' ');
-	numseq('actconn', 'feconn', 'beconn', 'srv_conn', 'retries');
+	group(CLS_DIGITS, 'actconn', 'number');
+	non('/');
+	group(CLS_DIGITS, 'feconn', 'number');
+	non('/');
+	group(CLS_DIGITS, 'beconn', 'number');
+	non('/');
+	group(CLS_DIGITS, 'srv_conn', 'number');
+	non('/');
+	group('\\+?', 'retries_redispatch', 'notempty');
+	group(CLS_DIGITS, 'retries', 'number');
+
 	non(' ');
-	numseq('srv_queue', 'backend_queue');
+	group(CLS_DIGITS, 'srv_queue', 'number');
+	non('/');
+	group(CLS_DIGITS, 'backend_queue', 'number');
+
 	non(' "');
 	group('[^"]*', 'http_request');
 	non('"$');
@@ -187,7 +202,9 @@ HAProxyLogTransform.prototype._transform = function (l, _, done) {
 	var out = {};
 
 	if (!m) {
-		setImmediate(done, VE('malformed log line: "%s"', l));
+		var re = self.hlt_matcher.re;
+		setImmediate(done,
+		    VE('malformed log line: "%s" (does not match %s)', l, re));
 		return;
 	}
 
@@ -211,6 +228,9 @@ HAProxyLogTransform.prototype._transform = function (l, _, done) {
 		case 'ip':
 			out[g.n] = normalise_ip(v);
 			break;
+		case 'notempty':
+			out[g.n] = (v !== '');
+			break;
 		default:
 			out[g.n] = m[g.i];
 			break;
diff --git a/package.json b/package.json
index c10fc95..f80ee2b 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
 	"name": "haproxy-log",
 	"description": "very basic haproxy log parser",
-	"version": "1.0.0",
+	"version": "1.0.1",
 	"main": "./lib/haproxy-log.js",
 	"bin": {
 	    "haplog": "./bin/haplog"
